### PR TITLE
Add support for swapping out levels

### PR DESCRIPTION
**Pull request by balderdash**
This PR modifies ScriptDSL to support entries like:

```
stage 'blah'
level 'foo'
variants
  level 'bar', active: true
  level 'baz', active: false
endvariants
level 'qux'
```

The `variants` section creates a script_level with two levels, one of which is indicated as inactive. The `active` bit is stored in a JSON blob in the new properties column in script_level.

I also had to modify the script_level, api, and activities controllers to serve the right level and store progress correctly.

### CODE REVIEW

File: dashboard/app/controllers/activities_controller.rb [hunk 1]
```diff
@@ -24,7 +24,7 @@ def milestone
 
     if params[:script_level_id]
       @script_level = ScriptLevel.cache_find(params[:script_level_id].to_i)
-      @level = @script_level.level
+      @level = (params[:level_id] && Level.find(params[:level_id].to_i)) || @script_level.levels[0]
```
Comment (Hamms) [inline dashboard/app/controllers/activities_controller.rb]: nit: a ternary operator would probably be clearer here, but that's arguably a matter of personal preference.

File: dashboard/app/controllers/api_controller.rb [hunk 2]
```diff
@@ -126,7 +126,7 @@ def section_text_responses
     load_section
     load_script
 
-    text_response_script_levels = @script.script_levels.includes(:levels).where('levels.type' => [TextMatch, FreeResponse])
+    text_response_script_levels = @script.script_levels.includes(:levels).where('levels.type' => TextMatch)
```
Comment (Hamms) [inline dashboard/app/controllers/api_controller.rb]: What's the reason for this change?

File: dashboard/app/controllers/script_levels_controller.rb [hunk 3]
```diff
@@ -199,9 +199,31 @@ def load_section
...
+      last_attempt = current_user.last_attempt_for_any(@script_level.levels)
+      return last_attempt.level if last_attempt
+    end
+
+    # Otherwise return the active level with the lowest id. Active defaults to true
+    return @script_level.levels.min_by(&:id) unless @script_level.properties
+
+    properties = JSON.parse(@script_level.properties)
+    @script_level.levels.sort_by(&:id).each do |level|
+      return level unless properties[level.name] && properties[level.name]['active'] == false
```
Comment (Hamms) [inline dashboard/app/controllers/script_levels_controller.rb]: you can more clearly accomplish this using [detect/find](http://apidock.com/ruby/Enumerable/find)

Comment (Hamms) [inline dashboard/app/controllers/script_levels_controller.rb]: It's also not clear to me that `id` is the best attribute to sort by; the incremental nature of ids is a side effect of one particular method of enforcing uniqueness, not a feature in and of itself. Creation time might be what we really want here.

File: code-studio/src/js/initApp/loadApp.js [hunk 4]
```diff
@@ -51,7 +51,12 @@ module.exports = function (callback) {
       appOptions.disableSocialShare = true;
     }
 
-    $.ajax('/api/user_progress/' + appOptions.scriptName + '/' + appOptions.stagePosition + '/' + appOptions.levelPosition).done(function (data) {
+    $.ajax(
+        '/api/user_progress/' + appOptions.scriptName +
+        '/' + appOptions.stagePosition +
+        '/' + appOptions.levelPosition +
+        '/' + appOptions.serverLevelId
```
Comment (Hamms) [inline code-studio/src/js/initApp/loadApp.js]: We can now do this with [template strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)!

File: dashboard/app/models/script.rb [hunk 5]
```diff
@@ -401,71 +401,82 @@ def self.add_script(options, data)
...
-    script.script_levels = data.map do |row|
-      row.symbolize_keys!
+    script.script_levels = raw_script_levels.map do |raw_script_level|
+      raw_script_level.symbolize_keys!
 
-      # Concepts are comma-separated, indexed by name
-      row[:concept_ids] = (concepts = row.delete(:concepts)) && concepts.split(',').map(&:strip).map do |concept_name|
-        (Concept.by_name(concept_name) || raise("missing concept '#{concept_name}'"))
-      end
+      assessment = false
```
Comment (Hamms) [inline dashboard/app/models/script.rb]: why `false` and not `nil`?

File: dashboard/app/dsl/script_dsl.rb [hunk 6]
```diff
@@ -76,21 +78,43 @@ def level_concept_difficulty(json)
...
       :skin => @skin,
       :concepts => @concepts.join(','),
       :level_concept_difficulty => @level_concept_difficulty || {},
       :video_key => @video_key_for_next_level
     }.merge(properties).select{|_, v| v.present? }
     @video_key_for_next_level = nil
+    if @current_scriptlevel
+      @current_scriptlevel[:levels] << level
+      if !@active
+        @current_scriptlevel[:properties][name] = { :active => @active }
```
Comment (Hamms) [inline dashboard/app/dsl/script_dsl.rb]: I'm curious, why add a properties object, when (unless I'm mistaken) all you're storing on it is an `active` bool? Why not just store the bool directly?

Comment (balderdash) [inline dashboard/app/dsl/script_dsl.rb]: I plan to add more properties later (see https://github.com/code-dot-org/code-dot-org/pull/8330), and I don't want to add a column for all of them.

File: dashboard/app/controllers/activities_controller.rb (hunk 1 above)
Comment (Hamms) [inline dashboard/app/controllers/activities_controller.rb]: actual note; do we really want to default to the first level, or do we want to be more clever?

We could move that `select_level` method you defined below into the ScriptLevel model and call it something a bit more verbose like `get_first_active_level`, then just call it here and in the other places we care about getting the right level.

File: dashboard/app/controllers/application_controller.rb [hunk 7]
```diff
@@ -120,10 +120,11 @@ def milestone_response(options)
       timestamp: DateTime.now.to_milliseconds
     }
     script_level = options[:script_level]
+    level = options[:level]
```
Comment (Hamms) [inline dashboard/app/controllers/application_controller.rb:123]: how easy is it for us to get into a situation where level is undefined?

File: dashboard/app/controllers/script_levels_controller.rb [hunk 8]
```diff
@@ -182,7 +183,6 @@ def load_user
     # TODO: This should use cancan/authorize.
     if user.student_of?(current_user)
       @user = user
-      @user_level = @user.user_level_for(@script_level)
```
Comment (Hamms) [inline dashboard/app/controllers/script_levels_controller.rb:185]: It's not clear to me why it's safe to move this declaration into a logically very distinct part of `load_level_source`; can you explain?

File: dashboard/config/routes.rb [hunk 9]
```diff
@@ -327,6 +328,7 @@ def redirect_to_teacher_dashboard
   get '/api/student_progress/:section_id/:student_id', to: 'api#student_progress', as: 'student_progress'
   get '/api/user_progress/:script_name', to: 'api#user_progress', as: 'user_progress'
   get '/api/user_progress/:script_name/:stage_position/:level_position', to: 'api#user_progress_for_stage', as: 'user_progress_for_stage'
+  get '/api/user_progress/:script_name/:stage_position/:level_position/:level', to: 'api#user_progress_for_stage', as: 'user_progress_for_stage_and_level'
```
Comment (Hamms) [inline dashboard/config/routes.rb:340]: are we actually using this anywhere?

**Comment by Hamms** [comment]
Finished with my first pass; it's looking good so far, but I have a handful of questions.

It would be good to get someone who has a little more familiarity than I to look into the DSL parsing code, but I'm afraid I'm not sure who the best candidate is.

Do we have a sense of whether or not this will have performance implications for level loading or progress/milstone posting? Even something as basic as comparing the number of SQL queries generated by accessing the various routes before and after this change would be awesome!

File: dashboard/app/controllers/activities_controller.rb (hunk 1 above)
Comment (balderdash) [inline dashboard/app/controllers/activities_controller.rb]: I switched to `oldest_active_level` here, but once this change is rolled out milestone posts will always include a level_id param. This is just a fallback in case I missed something, and to cover people who post progress right after the push.

File: dashboard/app/controllers/api_controller.rb (hunk 2 above)
Comment (balderdash) [inline dashboard/app/controllers/api_controller.rb]: Merge gone wrong, thanks for catching it.

File: dashboard/app/controllers/script_levels_controller.rb (hunk 3 above)
Comment (balderdash) [inline dashboard/app/controllers/script_levels_controller.rb]: Switched to find and created_at.

File: code-studio/src/js/initApp/loadApp.js (hunk 4 above)
Comment (balderdash) [inline code-studio/src/js/initApp/loadApp.js]: Switched, but still needed line breaks :(

File: dashboard/app/models/script.rb (hunk 5 above)
Comment (balderdash) [inline dashboard/app/models/script.rb]: Hmm, I saw that it was a boolean so I figured it should default to false, but it looks like the existing behavior sets it to `true` or `nil`. Switched.

File: dashboard/app/controllers/application_controller.rb (hunk 7 above)
Comment (balderdash) [inline dashboard/app/controllers/application_controller.rb:123]: This method is just called from the activities and script_level controllers, and both of those default to something sane (or explicitly raise an exception).

File: dashboard/app/controllers/script_levels_controller.rb (hunk 8 above)
Comment (balderdash) [inline dashboard/app/controllers/script_levels_controller.rb:185]: `@user_level` is only used in rendering the `levels/teacher` partial, when the teacher is viewing a student's solution. `load_level_source` is always called before rendering `levels/show`, so it's safe to set `@user_level` there instead, and the section I moved it to is for handling the "teacher viewing student's solution" case as well.

I moved it because I now need to determine which level to render before setting `@user_level`.

File: dashboard/config/routes.rb (hunk 9 above)
Comment (balderdash) [inline dashboard/config/routes.rb:340]: Yep, that was the change in initApp/loadApp.js. If we don't specify `:level`, we end up showing the wrong levelsource to students who attempt one version of a level and then load another.

**Comment by balderdash** [comment]
Re: performance

For the most common scenario (one level per scriptlevel), there should be zero additional queries, and marginal changes overall (I'm adding a couple checks of scriptlevel.levels.length, but scriptlevel.levels is cached).

That's confirmed by my local testing. Level loads take zero queries both before and after this change. Milestone posts take 2-3 queries both before and after (levelsource select and sometimes insert, and levelsourcehint select).

File: dashboard/app/dsl/script_dsl.rb (hunk 6 above)
Comment (wjordan) [inline dashboard/app/dsl/script_dsl.rb]: You might want to take a look at the [`SerializedProperties`](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/concerns/serialized_properties.rb) concern, which is what we currently use to apply this pattern to the [`Level`](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/level.rb#L48-L57) and [`User`](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/user.rb#L78-L79) models. I'd recommend reusing this Concern if possible since it reduces the amount of (de)serialization code scattered through the implementation.

Comment (wjordan) [inline dashboard/app/dsl/script_dsl.rb]: Or actually, [`Store`](http://api.rubyonrails.org/classes/ActiveRecord/Store.html) might be the more 'official' way to apply pattern in Rails (I didn't know about `Store` when I initially wrote SerializedProperties, but it might be worth using `Store` for new/future applications of this pattern).

**Comment by balderdash** [comment]
@breville Hey Brendan, can you take a look at the DSL parsing changes in this PR? app/dsl/script_dsl.rb and app/models/script.rb

Edit: nvm, @wjordan is on it.

File: dashboard/app/dsl/script_dsl.rb [hunk 10]
```diff
@@ -76,21 +78,43 @@ def level_concept_difficulty(json)
...
+    end
+  end
+
+  def variants
+    @current_scriptlevel = { :levels => [], :properties => {}, :stage => @stage}
+  end
+
+  def endvariants
+    @scriptlevels << @current_scriptlevel
+    @current_scriptlevel = nil
```
Comment (wjordan) [inline dashboard/app/dsl/script_dsl.rb:128]: I'm wondering if it's worth using a more conventional DSL format here by replacing `variants`/`endvariants` methods with a single block-based method `variants(&block)`. For example:

``` ruby
stage 'Stage1'
level 'Level 1'
variants {
  level 'Level 2a'
  active false
  level 'Level 2b'
}
level 'Level 3'
```

though if we really wanted to do this 'right', we should replace all other stateful methods like `stage` with block parameters as well, e.g.:

``` ruby
stage 'Stage1' {
  level 'Level1'
  variants {
    level 'Level 2a'
    level 'Level 2b', active: false
  }
}
```

For this reason maybe it's worth making such a change later as a bigger refactor, since we have a bunch of existing stateful methods like this that we would want to change (and mass-migrate all of our existing script DSL files to) for consistency sake.

@joshlory thoughts? I know in the past we had discussed possibly making the Script DSL obsolete at some point anyway, so maybe not even worth investing even more effort into a clean design here?

Comment (Hamms) [inline dashboard/app/dsl/script_dsl.rb:128]: I love that proposal for a block-based format. Is the issue that we would have to add support for block-based methods in general, or simply that it would be nice to have consistency?

Comment (joshlory) [inline dashboard/app/dsl/script_dsl.rb:128]: As we add more concepts to the `.script` format I think we're going to want to eventually create a richer editor than just text.  I don't have a good sense of when the cutoff point is for switching formats (maybe we've already passed it).

If we do go the block route, we should consider moving some existing flat options to nested options.  E.g. all the [metadata for legacy levels](https://github.com/code-dot-org/code-dot-org/blob/e389815a37c0a2eaefbddc9a00685a197f678bd7/dashboard/config/scripts/20-hour.script#L90-L93):

``` ruby
...
skin 'pvz'
concepts 'if_else', 'loop_until'
level_concept_difficulty '{"sequencing":5,"repeat_until_while":4,"conditionals":4}'
level 'blockly:Maze:2_18'

skin 'pvz'
concepts 'if_else', 'loop_until'
level_concept_difficulty '{"sequencing":5,"repeat_until_while":4,"conditionals":5}'
level 'blockly:Maze:2_19'
...
```

**Comment by breville** [comment]
Looks like the right people have already seen this.

could we at least squish this:

  active true
  level 'bar'

into

  level 'bar', active: true

i think that would be more readable, short of doing the more elaborate block change.

**Comment by Hamms** [comment]
Do note that https://github.com/code-dot-org/code-dot-org/pull/8330 adds some more parameters; I'm not sure we'll want to squish when we have more than just `active`

**Comment by breville** [comment]
Even then, can we wrap the parameters in the DSL like this?

```
level 'bar', 
  active: true,
  buttontext 'Challenge',
  imageurl 'https://studio.code.org/blah/maze-2-challenge.png',
  description 'This is a hard level'

```

I feel like that'd be a bit easier to read.

**Comment by balderdash** [comment]
I do like that better, updated script_dsl.rb accordingly.

**Comment by Hamms** [comment]
LGTM!
